### PR TITLE
Tuple/List get_item to fallible

### PR DESCRIPTION
**Pull request by aviramha**
Hey, trying to help with:
https://github.com/PyO3/pyo3/issues/1667
specifically:
make PyTuple::get_item and PyList::get_item fallible
I'm not quite sure about the extract thingy, any guidance or feedback will be welcome.

### CODE REVIEW

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:479]: 👍 good idea to add a test, it could be nice to also check the error type (`IndexError`?) or message?

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:141]: I think it's fine to use `.expect` here.


```suggestion
            let item = self.tuple.get_item(self.index).expect("index is less than length");;
            self.index += 1;
            Some(item)
```

If you don't feel comfortable about the possble panic, it would also be possible to implement a `#[cfg(not(Py_LIMITED_API))]` version of `PyTupleIterator` which uses `as_slice`.

**Review by davidhewitt — COMMENTED** [review]
Thanks for kicking this off!

Another way to design this which is potentially for backwards compatibility is to deprecate `.get_item()` and to introduce two new methods:
 - `.get() -> PyResult<&PyAny>` 
 - `.get_unchecked() -> &PyAny`, doesn't do checking and only available on `#[cfg(not(Py_LIMITED_API))]`

Or `.get() -> Option<&PyAny>`, if handling the error internally feels appropriate. I'd be open to opinions on this.

Regarding the `.extract()` error, you should check for uses of `.get_item()` inside `pyo3-macros-backend`. TBH this is an example why deprecating `get_item` might be the better idea - the compilation will still go through and you'll just get warnings of where it's used, which can be fixed instead.

File: src/types/tuple.rs
Comment (aviramha) [reply src/types/tuple.rs:141]: Overall I'm against having safe code that can panic (even if it's induced by very odd human error)
I'll check the second suggestion.

**Comment by aviramha** [comment]
> Or .get() -> Option<&PyAny>, if handling the error internally feels appropriate. I'd be open to opinions on this.

I like this better. I don't like that the user might panic when using the API (while the API is safe).

**Comment by aviramha** [comment]
@davidhewitt I've changed - I wasn't sure if to handle the deprecations now or maybe later?

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:97]: This one is definitely `unsafe`:

```suggestion
    /// Gets the list item at the specified index. Undefined behavior on bad index. Use with caution.
    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
    pub unsafe fn get_unchecked(&self, index: usize) -> &PyAny {
```

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:450]: Can we use `Python::with_gil` in new tests please :)

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:79]:
```suggestion
    #[deprecated = "use PyList::get, which does not panic"]
```

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:145]: This can use `get_unchecked` when available.

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:222]: Same here r.e. `get_unchecked`.

**Review by davidhewitt — COMMENTED** [review]
Thanks, this is taking shape!

@birkenfeld as you came up with the proposals in #1667 I'd be really interested for your opinion on this.

**Review by mejrs — COMMENTED** [review]
Thanks :)

I do have some suggestions..

File: src/types/list.rs
Comment (aviramha) [reply src/types/list.rs:88]: We want negative indexing, but when using index (slicing). There's no real sense in implementing it at this function level.

Comment (mejrs) [reply src/types/list.rs:97]: Personally I'm leaning towards using usize everywhere, anyway 👍

**Review by mejrs — APPROVED** [review]

File: src/types/list.rs
Comment (birkenfeld) [reply src/types/list.rs:97]: I don't see why this should only be relevant for slicing. Python allows negative indices for single item access as well as for slices.  IMO either we allow negative indices for both, or none.

File: src/types/list.rs
Comment (birkenfeld) [inline src/types/list.rs:184]: It's a list here

**Review by birkenfeld — CHANGES_REQUESTED** [review]
I'm happy with returning `PyResult`.  The new name makes the API asymmetric (`get` vs `set_item`, and btw why is there is no `del_item` or `remove`...)  Are we still allowed to break APIs if the result is a clear compile error?

`usize` vs `isize` needs to be discussed still.

A follow-up to this is the implementation of `Index` traits.

**Comment by davidhewitt** [comment]
> The new name makes the API asymmetric (get vs set_item, and btw why is there is no del_item or remove...) Are we still allowed to break APIs if the result is a clear compile error?

Yes, we could change `get_item` in 0.15. I do agree we should keep consistent names; maybe it is better to have `get_item` and `get_item_unchecked` here. I think I prefer having `isize` because it's more similar to the Python semantics, but I could be persuaded either way. The underlying C-API only accepts unsigned integers, so we'd need to do the conversion (but that's not _too_ bad)...

Sorry @aviramha you picked a ticket which has a lot of open design questions 🙃 

I definitely think if we call it `.get()` then we should accept `usize` only.

**Comment by aviramha** [comment]
> >  Sorry @aviramha you picked a ticket which has a lot of open design questions 🙃

That's OK. the get_item of Python is actually usize, so it might be more confusing that it works with negative indices. I guess it's a question whether you want the API to feel like a Rust wrapper to the CPython or a rust wrapper to the Python API.
Let me know which you prefer and I'll fix the PR.

**Comment by davidhewitt** [comment]
That's a fair question. I think I'm now leaning towards this matching the C API, so would vote for `get_item(usize) -> PyResult<&PyAny>` and `unsafe get_item_unchecked(usize) -> &PyAny`.

**Comment by aviramha** [comment]
> That's a fair question. I think I'm now leaning towards this matching the C API, so would vote for `get_item(usize) -> PyResult<&PyAny>` and `unsafe get_item_unchecked(usize) -> &PyAny`.

Goes without saying that I am too. How do we make this call? 🙃

**Comment by davidhewitt** [comment]
If it looks like a majority of us agree, I'm happy for us to commit to this plan (cc @birkenfeld @mejrs @messense).

**Comment by birkenfeld** [comment]
Where would `isize` be allowed then? This should really be designed keeping all places with indices in mind (see #1667), not decided in a case-by-case basis just looking at the C API, which is inconsistent. For example, `PyList_Insert` accepts negative indices.

**Comment by birkenfeld** [comment]
To be clear, I'm not opposed to doing away with `isize` everywhere and only accepting "real" indices, Rust-std-style.  I just want it to be consistent.

**Comment by aviramha** [comment]
@birkenfeld I understand your concern but I believe that's CPython inconsistency issue. It depends on what we value more - being consistent with it or improving it.

**Comment by aviramha** [comment]
To make myself more clear (previous comment was from mobile) - when coding in PyO3 I often use the CPython docs to understand how stuff works - having different API (i.e get_item of PyO3 accepts negative indices vs CPython not) can easily cause confusion for anyone.

**Comment by birkenfeld** [comment]
The expectation of each PyO3 method mapping exactly to a Python C-API call is misguided, though.  If you are missing info from the PyO3 docs which is integral to understanding the API in question, it is a bug and should be fixed; please report that here.

And of course, in the case we're discussing here, the documentation needs to state that negative indices are supported and how they are interpreted. (That this is not documented currently is one of the points of #1667).

**Comment by davidhewitt** [comment]
I definitely agree that we should go for all-`isize`, or all-`usize`. It's not fun as a user to have to look up argument types for every function. (And the right document source should definitely be PyO3 docs, not CPython docs.)

Here's the arguments I can see either way:

---
#### usize
Good:
  - Matches `std`
  - Means we can use the `SliceIndex` trait to allow doing `list[0..5]` etc

Bad:
  - `PySequence` and `PyList_Insert` APIs work correctly with negative indices, but we'd be preventing users from doing so.

---
#### isize
Good:
  - Matches python usage, e.g. `list[-5]`
  - We'd have to work around APIs that don't support negative indexing. It's probably as simple as using `PySequence_GetItem` instead of `PyList_GetItem`, and similar, but not sure. Some edge cases might be annoying.

Bad:
   - Supporting negative indexing presumably has a (tiny) performance overhead, because negative indices must be converted to actual positive ones.
   - To have slices with negative indices such as `list[-4..-2]` we'd probably need to create a `PySliceIndex` trait which is implemented for various `Range<isize>`, `RangeFrom<isize>`, `RangeFull<isize>`. Could be a lot of work.
   - `PyList::get_item_unchecked` would almost certainly want to take `usize`. Either that or it could have a simple check `if idx < self.len() { idx += self.len() }`, But that would again imply performance overhead, which kinda sucks in an `unsafe` low-level API.

---

An alternative... what if we made `PyList` take `usize` and `PySequence` take `isize`? That way we'd get the simplicity of `usize` for most usage but also give users the "full API" if they chose to treat the list as a sequence. We'd have to document this really clearly, but at least there's then a clear pattern to it.

I think this "mixed-mode" would match the C-API in every way except `PyList::insert`, which we would make only take `usize`.

**Comment by mejrs** [comment]
@davidhewitt thanks for the recap

Indexing and slicing with negative numbers doesn't seem like a good idea to me:
- For someone not coming from Python, having negative indices work when indexing/slicing is something that I feel will mostly just be surprising and unexpected.
- For someone who *is* coming from Python (to Rust), I think it'll be confusing if negative indexing works somewhere (pyo3) but not in other places (the rest of the Rust ecosystem).
- It will only be a convenience for people making Python libraries rather than the people using said libraries. After all, authors can implement negative slicing and indexing for their own types.
- I'm not a fan of doing a bunch of math tricks in what should be a low level api.

**Comment by davidhewitt** [comment]
Agreed, I'm now feeling very sure that `usize` is definitely the correct choice for `PyList`.

The question for me then becomes whether `PySequence`should be based around `isize` to match the underlying Python APIs. On balance, I think I slightly err towards `PuSequence` using `usize` too for simplicity.

**Comment by aviramha** [comment]
@davidhewitt can I finish the PR then? :) (finalizing the changes)

**Comment by davidhewitt** [comment]
Sure thing, let's go with the `usize` plan!

**Comment by birkenfeld** [comment]
Agreed!

**Comment by aviramha** [comment]
@davidhewitt @birkenfeld should I include anything else then in this PR?

**Comment by davidhewitt** [comment]
I understood we decided above that we should have `get_item(&self, usize) -> PyResult<&PyAny>` and `unsafe.get_item_unchecked(&self, usize) -> &PyAny`?

The `.get()` suggestion by me will need to be rolled back - sorry about that.

This also will need a CHANGELOG entry.

In addition, if we're changing `PyList` and `PyTuple` APIs to take `usize` in this PR it would be ideal if other relevant APIs were updated at the same time. (Although this is not entirely necessary.)

I'm about to release 0.14.2 tomorrow, and happy to see this merged after that.

**Comment by aviramha** [comment]
@davidhewitt should I change get_slice to also use usize?

**Comment by aviramha** [comment]
I'm not sure how I broke the extract thingy

**Comment by davidhewitt** [comment]
Check for uses of .get_item inside `pyo3-macros-backend`. I expect code is being generated which used to just use the value, but now there's results being returned instead.

**Comment by aviramha** [comment]
Cool. @davidhewitt Waiting for your review :

**Comment by davidhewitt** [comment]
👍 will try to give a full read within the next few days!

File: CHANGELOG.md
Comment (davidhewitt) [inline CHANGELOG.md:32]: Afraid since we released 0.14.2 you'll need to rebase on `main` and move this CHANGELOG entry up to the new "Unreleased" section building for 0.15.

I think for users the return type is more relevant, needs PR number, and I'd mention `get_item_unchecked` separately:

```suggestion
### Added
- Add `PyList::get_item_unchecked()` and `PyTuple::get_item_unchecked()` to get items without bounds checks. [#1733](https://github.com/PyO3/pyo3/pull/1733)

### Changed 
- `PyList` and `PyTuple`'s `get_item` now return `PyResult<&PyAny> instead of panicking. [#1733](https://github.com/PyO3/pyo3/pull/1733)
```

File: benches/bench_list.rs
Comment (davidhewitt) [inline benches/bench_list.rs:35]: It might be nice to add a `list_get_item_unchecked` benchmark.

File: benches/bench_tuple.rs
Comment (davidhewitt) [inline benches/bench_tuple.rs:35]: Similarly it might be nice to add a `tuple_get_item_unchecked` benchmark.

File: pyo3-macros-backend/src/from_pyobject.rs
Comment (davidhewitt) [inline pyo3-macros-backend/src/from_pyobject.rs:246]: I think this probably needs a `.and_then` or something rather than returning early, because we still need the error message:

```suggestion
                s.get_item(#i).and_then(PyAny::extract).map_err(|inner| {
```

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:82]: The original code incremented the reference and used `from_owned_ptr()`.

That's still necessary here - the problem with `from_borrowed_ptr` is that if a user then edits the `PyList` the "borrowed" reference could be deleted out from underneath us.

See #890 for more detail.

```suggestion
            // PyList_GetItem return borrowed ptr; must make owned for safety (see #890).
            ffi::Py_INCREF(item);
            self.py().from_owned_ptr(item)
```

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:94]: Same again here:

```suggestion
        // PyList_GET_ITEM return borrowed ptr; must make owned for safety (see #890).
        ffi::Py_INCREF(item);
        self.py().from_owned_ptr(item)
```

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:156]: I think with this change there's some conversions in `size_hint` (`self.index as usize`) which aren't needed any more.

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:226]: Nit: bonus newline

```suggestion
```

**Review by davidhewitt — CHANGES_REQUESTED** [review]
Thanks for pushing forward on this! I'm afraid there's a few bits and pieces which still need tidying up. Comments scattered below.

> @davidhewitt should I change get_slice to also use usize?

Before we release 0.15 we'll need to change all these tuple / list / sequence APIs which currently take `isize` to take `usize`. This PR is already big enough, so maybe it's better to make a list of what still needs to change in #1667 and then make follow-up PRs.

**Comment by aviramha** [comment]
Thanks for the review. I think I fixed all of the comments :)

File: src/types/tuple.rs
Comment (davidhewitt) [inline src/types/tuple.rs:100]: Similarly here:

```suggestion
    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
    #[cfg_attr(docsrs, doc(cfg(not(any(Py_LIMITED_API, PyPy))))]
```

**Review by davidhewitt — COMMENTED** [review]
The changes are looking reasonable to me... but CI has a segfault?

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:483]: Ah I guess this test is just not worth having - you're reading an invalid memory location. No wonder it crashes!

**Comment by aviramha** [comment]
Oopsie, I thought the CI failure was main's fault

File: src/types/list.rs
Comment (davidhewitt) [inline src/types/list.rs:83]: Ahh, the segfault on CI is my fault. When there's an error item is null. Best way to handle that is use `Py_XINCREF`:

```suggestion
            ffi::Py_XINCREF(item);
```

**Review by davidhewitt — COMMENTED** [review]
Ahh I think I've found another cause of segfault... Fingers crossed CI green after 🤞

**Review by davidhewitt — APPROVED** [review]
👍 looks brilliant! Thanks again for your patience and many iterations as we worked out the design for this. 

I'll merge this shortly (just going to cherry-pick some fixes for a 0.14.3 bugfix release first).

**Comment by davidhewitt** [comment]
(See #1796 - as soon as that release is live I'll merge this. It just makes my life easier if I try to cherry-pick other stuff and there's no more breaking changes 😉)

**Review by birkenfeld — APPROVED** [review]
LGTM.

**Comment by aviramha** [comment]
> 👍 looks brilliant! Thanks again for your patience and many iterations as we worked out the design for this. 
> 
> 
> 
> I'll merge this shortly (just going to cherry-pick some fixes for a 0.14.3 bugfix release first).

You're welcome! Glad to contribute to an awesome framework.
Where do we wish to track progress on this so I might pick up the next task?

**Comment by davidhewitt** [comment]
> You're welcome! Glad to contribute to an awesome framework.
Where do we wish to track progress on this so I might pick up the next task?

🚀

If you want to own this, that would be awesome! The easiest way might be to create a new issue where you can own a list of checkboxes and close them as we make progress?

**Comment by birkenfeld** [comment]
Since 0.14.x is going to be in a branch, can we merge this?

**Comment by davidhewitt** [comment]
Ok 👍 I'm getting more used to doing the cherry-picking onto the 0.14.3 branch now, so let's proceed with this.

I see I created a merge conflict by merging #1802, so I'll fixup this quickly now...

**Pull request merged**
